feat: Discard transactions

rutwikhdev opened the pull request:
Fixes: https://github.com/frappe/frappe/issues/21515

Added discard action under menu items on draft transactions. This action will do transition validation, check write perms, run hooks and do db_set with docstatus 2.

![image](https://github.com/frappe/frappe/assets/50401596/d5a76aa4-ac64-469a-a57e-c0aff1469550)

Added Events
---
Override doc methods: before_discard, on_discard
Client Script: before_discard, after_discard (following same naming conventions as cancel)
Server Script Doctype Events: Before Discard, After Discard
 
Right now it just validates write perms, perhaps we can add separate permissions later.


docs: https://frappeframework.com/docs/user/en/api/form#form-events

rutwikhdev committed:
feat: discard draft transactions

rutwikhdev committed:
chore: add discard events for server script

rutwikhdev committed:
test: add test for discard action

ankush commented on frappe/tests/test_document.py:
```diff
@@ -98,6 +98,13 @@ def test_update(self):
 
 		self.assertEqual(frappe.db.get_value(d.doctype, d.name, "subject"), "subject changed")
 
+	def test_discard(self):
+		d = self.test_insert()
+		self.assertEqual(d.docstatus, 0)
+
+		d.discard()
+		self.assertEqual(d.docstatus, 2)
+
```
ummm, add more tests. Think of all the use cases / transitions you're promising with this feature.

ankush replied:
- cancel/submit -> discard shouldn't work
- discard to draft shouldn't work
- discard shouldn't trigger any on_cancel events
- discard shouldn't trigger save of entire doc, if any code does it from events it should cause a validation failure (cause save might endup triggering cancel events)
- `_action=discard` should be asserted and usage be checked in code

ankush replied:
This PR will certainly require more tests than actual production code

rutwikhdev replied:
We can drop the _action change because I'm doing transition checks explicitly in discard

```python		
if not self.docstatus == DocStatus.draft():
    raise frappe.ValidationError(_("Only draft documents can be discarded"), self.docstatus)
```

Doing this because of some conflicting rules that should be invalid in other scenarios like,
0 > 2 - should not be allowed for draft.cancel() but allowed for draft.discard()
1 > 2 - should not be allowed for submitted.discard() but allowed for submitted.cancel()

ankush commented on frappe/model/document.py:
```diff
@@ -848,9 +848,7 @@ def check_docstatus_transition(self, to_docstatus):
-				raise frappe.DocstatusTransitionError(
-					_("Cannot change docstatus from 0 (Draft) to 2 (Cancelled)")
-				)
+				self.check_permission("cancel")
```
`draft->discard` should just require `write` instead of `cancel`. Cancel perm is for cancelling submitted docs.

ankush commented on frappe/desk/form/save.py:
```diff
@@ -59,6 +59,17 @@ def cancel(doctype=None, name=None, workflow_state_fieldname=None, workflow_stat
 
+@frappe.whitelist()
+def discard(doctype=None, name=None):
+	"""discard a doclist"""
```
misleading/old docstring

ankush commented on frappe/desk/form/save.py:
```diff
@@ -59,6 +59,17 @@ def cancel(doctype=None, name=None, workflow_state_fieldname=None, workflow_stat
 
 
+@frappe.whitelist()
+def discard(doctype=None, name=None):
```
```suggestion
def discard(doctype: str, name: str | int):
```

Why none defaults? Don't tolerate bad input that is impossible to handle in APIs.

rutwikhdev replied:
Friendship cancelled with co-pilot

ankush commented on frappe/public/js/frappe/form/form.js:
```diff
@@ -833,6 +833,18 @@ frappe.ui.form.Form = class FrappeForm {
 	}
 
+	discard(btn, callback, on_error) {
+		var me = this;
```
```suggestion
		let me = this;
```

upgrade old js if copying :smile:

ankush commented on frappe/public/js/frappe/form/save.js:
```diff
@@ -274,6 +274,23 @@ frappe.ui.form.save = function (frm, action, callback, btn) {
 	}
 };
 
+frappe.ui.form.discard = function (frm, callback, btn) {
```
Don't copy old bad patterns. This code should be on form only and not bolted on to global namespaces created using `frappe.provide`

ankush commented:
TODO: 

- [x] behavior when workflow is active (?)

rutwikhdev committed:
fix: misc fixes for discard action

* use write perms instead of cancel
* update docstring
* remove discard from global namespace

rutwikhdev committed:
chore: revert transition rule for 0 to 2

doing explicit transition check for discard because,
* there's only one transition check that is required
* draft(0) > cancelled(2) and submitted(1) > cancelled(2) are valid
  checkes for save so it doesn't make sense editing
  check_docstatus_transition

rutwikhdev committed:
test: add more tests for discard

ankush commented on frappe/desk/form/save.py:
```diff
@@ -60,8 +60,8 @@ def cancel(doctype=None, name=None, workflow_state_fieldname=None, workflow_stat
-	"""discard a doclist"""
+def discard(doctype: str, name: str | int):
+	"""dicard a draft document"""
 	doc = frappe.get_doc(doctype, name)
```
Typo.

rutwikhdev committed:
fix: typo

rutwikhdev committed:
chore: hide discard button if workflow is active and reuse
frappe.model.has_workflow

cogk commented:
Suggestion: maybe both the <kbd>Discard</kbd> and <kbd>Delete</kbd> menu items could be merged, with a popup to choose between the two. The advantage is that you have the space to explain the difference between discarding and deleting a draft document.

ankush commented:
@cogk they seem like 2 very different actions :thinking:

ankush commented:
Also in some cases we just hide delete based on perms or local requirements.

ankush commented on frappe/desk/form/save.py:
```diff
@@ -59,6 +59,17 @@ def cancel(doctype=None, name=None, workflow_state_fieldname=None, workflow_stat
+	doc = frappe.get_doc(doctype, name)
+	capture_doc(doc, "Discard")
+
+	doc.discard()
```
One concern is clash with existing fieldnames/method :skull: 

Nothing in our codebase, but can happen in wild.

ankush commented on frappe/model/document.py:
```diff
@@ -1058,6 +1058,21 @@ def cancel(self):
+	def discard(self):
+		"""Discard the draft document. Sets `docstatus` = 2 with db_set."""
+		self.check_if_locked()
+		self.set_user_and_timestamp()
```
`check_if_latest` should be done here to avoid parallel updates. This has saved us from so many Heisenbugs.

ankush replied:
Doctype transition check might need some exception for this case somehow.

rutwikhdev replied:
@ankush done

ankush commented on frappe/public/js/frappe/form/form.js:
```diff
@@ -833,6 +833,18 @@ frappe.ui.form.Form = class FrappeForm {
+	discard(btn, callback, on_error) {
+		const me = this;
+		return new Promise((resolve) => {
+			// this.validate_form_action("Discard") // ?
```
```suggestion
			// this.validate_form_action("Discard") // ?
```

Drop dead comments, validation not not really required here IMO. Validation on other action helps because they are saving some local changes. 


Even client script hooks I'd say aren't required for this.

rutwikhdev committed:
chore: check_if_latest for discard action

rutwikhdev committed:
chore: drop dead comment
